### PR TITLE
Implement device functions to simplify writing kernel code

### PR DESCRIPTION
**Pull request by fwyzard**
Implement device functions to simplify writing kernel code:
  - `oncePerGrid(acc)`;
  - `oncePerBlock(acc)`;
  - `uniformElements(acc, size)` and `uniformElements(acc, begin, end)`;
  - `uniformGroups(acc, size)` and `uniformGroupElements(acc, group, size)`;
  - `independentGroups(acc, groups)` and `independentGroupElements(acc, group_size)`;
  - variations of these functions along different dimensions.

Implement tests for the most common functions.

### CODE REVIEW

File: include/alpaka/exec/Once.hpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    /* oncePerGrid
+     *
+     * `oncePerGrid(acc)` returns true for a single thread within the kernel execution grid.
+     *
+     * Usually the condition is true for block 0 and thread 0, but these indices should not be relied upon.
+     */
+
+    template<typename TAcc, typename = std::enable_if_t<isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC inline constexpr bool oncePerGrid(TAcc const& acc)
```
Comment (mehmetyusufoglu) [inline include/alpaka/exec/Once.hpp:26]: Maybe?
```suggestion
    ALPAKA_FN_ACC inline constexpr bool trueOncePerGrid(TAcc const& acc)
```

File: test/unit/exec/src/IndependentElements.cpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+#    pragma clang diagnostic ignored "-Wexit-time-destructors"
+#endif
+// Global Host object used by all tests.
+using Host = alpaka::DevCpu;
+static Host host = alpaka::getDevByIdx(alpaka::PlatformCpu{}, 0);
+#if BOOST_COMP_CLANG
+#    pragma clang diagnostic pop
+#endif
+
+/* Add the group id to te value of each element in the group.
```
Comment (mehmetyusufoglu) [inline test/unit/exec/src/IndependentElements.cpp]:
```suggestion
/* Add the group id to the value of each element in the group.
```

File: test/unit/exec/src/Once.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            ALPAKA_CHECK(*status, *value == 0);
+
+            alpaka::atomicAdd(acc, value, 1, alpaka::hierarchy::Grids{});
+
+            ALPAKA_CHECK(*status, *value == 1);
+        }
+    }
+};
+
+// The host and deivce compiler and libraries may use a different representation for boolean values. Reasonably enough,
```
Comment (mehmetyusufoglu) [inline test/unit/exec/src/Once.cpp]:
```suggestion
// The host and device compiler and libraries may use a different representation for boolean values. Reasonably enough,
```

File: test/unit/exec/src/IndependentElements.cpp [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+    // Initialise the accelerator platform.
+    Platform platform{};
+
+    // Random number generator with a gaussian distribution.
+    std::random_device rd{};
+    std::default_random_engine engine{rd()};
+
+    // Uniform distribution.
+    std::uniform_int_distribution<Idx> random_size{100, 201};
```
Comment (mehmetyusufoglu) [inline test/unit/exec/src/IndependentElements.cpp:93]:
```suggestion
    std::uniform_int_distribution<Idx> randomSize{100, 201};
```

File: test/unit/exec/src/IndependentElements.cpp [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    // Tolerance.
+    constexpr float epsilon = 0.000001f;
+
+    // Buffer size.
+    const Idx size = indices_h[groups];
+
+    // Allocate the input and output host buffer in pinned memory accessible by the Platform devices.
+    auto in_h = alpaka::allocMappedBuf<float, Idx>(host, platform, size);
```
Comment (mehmetyusufoglu) [inline test/unit/exec/src/IndependentElements.cpp:115]: Camel Case variable naming?
```suggestion
    auto inH = alpaka::allocMappedBuf<float, Idx>(host, platform, size);
```

File: include/alpaka/exec/Once.hpp (hunk 1 above)
Comment (fwyzard) [reply include/alpaka/exec/Once.hpp:26]: No.

File: test/unit/exec/src/IndependentElements.cpp (hunk 2 above)
Comment (fwyzard) [reply test/unit/exec/src/IndependentElements.cpp]: Will fix, thanks.

File: test/unit/exec/src/Once.cpp (hunk 3 above)
Comment (fwyzard) [reply test/unit/exec/src/Once.cpp]: Will fix, thanks.

File: test/unit/exec/src/IndependentElements.cpp (hunk 5 above)
Comment (fwyzard) [reply test/unit/exec/src/IndependentElements.cpp:115]: I don't particularly like this naming style.
I follow it for user-facing function names, like in the rest of alpaka, but I don't like to use it for variable names.

Comment (mehmetyusufoglu) [reply test/unit/exec/src/IndependentElements.cpp:115]: Yes, i also don't like actually.

**Comment by fwyzard** [comment]
let's see if the 23rd time's a charm...

**Comment by SimeonEhrig** [comment]
I check your code locally. The problem is the `AccCpuThreads` backend. On our workstation with 32 Core Epyc and Clang 17 debug build, the test already runs 100s. I think the annotation of the thread sanitizer also add a big overhead and the CI runners are weaker than our workstation.

I suggest to disable the tests with `std::thread` backend or if it make sense to test the thread backend, decrease the problem size.

**Comment by fwyzard** [comment]
> I suggest to disable the tests with `std::thread` backend or if it make sense to test the thread backend, decrease the problem size.

I agree... I've tried reducing the block size in all tests to 32 threads (or elements) per block.

**Comment by fwyzard** [comment]
Closed in favour of https://github.com/alpaka-group/alpaka/pull/2369 that includes further fixes and clean up.

**Pull request closed without merging**